src/Raven.Server/Documents/TimeSeries/TimeSeriesPolicyRunner.cs

```diff
@@ -0,0 +1,683 @@
+        {
+            private readonly TimeSeriesConfiguration _configuration;
+            private readonly List<RollUpState> _states;
+            private readonly bool _isFirstInTopology;
+
+            public long RolledUp;
+
+            internal RollupTimeSeriesCommand(TimeSeriesConfiguration configuration, List<RollUpState> states, bool isFirstInTopology)
+            {
+                _configuration = configuration;
+                _states = states;
+                _isFirstInTopology = isFirstInTopology;
+            }
+            protected override long ExecuteCmd(DocumentsOperationContext context)
+            {
+                var tss = context.DocumentDatabase.DocumentsStorage.TimeSeriesStorage;
+                RollUpSchema.Create(context.Transaction.InnerTransaction, TimeSeriesRollUps, 16);
+                var table = context.Transaction.InnerTransaction.OpenTable(RollUpSchema, TimeSeriesRollUps);
+                var toRemove = new List<Slice>();
+
+                foreach (var item in _states)
+                {
+                    if (_configuration == null)
+                        return RolledUp;
+
+                    var config = _configuration.Collections[item.Collection];
+                    if (config.Disabled)
+                        continue;
+                        
+                    var policy = config.GetPolicyByName(item.RollUpPolicy);
+                    if (policy == null)
+                        continue;
+
+                    if (table.ReadByKey(item.Key, out var current) == false)
+                    {
+                        toRemove.Add(item.Key);
+                        continue;
+                    }
+
+                    if (item.Etag != DocumentsStorage.TableValueToLong((int)RollUpColumn.Etag, ref current))
+                        continue; // concurrency check
+
+                    var startFrom = item.NextRollUp.Add(-policy.AggregationTime);
+                    var rawTimeSeries = item.Name.Split(TimeSeriesConfiguration.TimeSeriesRollupSeparator)[0];
+                    var intoTimeSeries = policy.GetTimeSeriesName(rawTimeSeries);
+                    
+                    var intoReader = tss.GetReader(context, item.DocId, intoTimeSeries, startFrom, startFrom);
+                    var previouslyAggregated = intoReader.AllValues().Any();
+                    if (previouslyAggregated)
+                    {
+                        var changeVector = intoReader.GetCurrentSegmentChangeVector();
+                        if (ChangeVectorUtils.GetConflictStatus(item.ChangeVector, changeVector) == ConflictStatus.AlreadyMerged)
+                        {
+                            // this rollup is already done
+                            toRemove.Add(item.Key);
+                            continue;
+                        }
+                    }
+
+                    if (_isFirstInTopology == false)
```

Comment:
Why do we have to get to this point to bail out? To remove the policies markers?

Reply:
Yes, remove rollups that were already done and replicated.